File: LCS_FINAL_JAVA_VERSION/Longest Common Substring/src/com/tsp/se/main/LongestCommonSubstring.java
package com.tsp.se.main;

import java.io.IOException;
import java.util.Scanner;

import com.tsp.se.inputOutput.FileManager;
import com.tsp.se.lcs.LongestCommonSubstringFinder;

/**
 * This Program's main objective is to take two text files and print on the
 * standard output the longest common substring between them.
 * <p>
 * At the beginning of the program it gives the user a choice between:
 * <ul>
 * <li>Giving the path of 2 already existing files
 * <li>Asking the program to generate n random files
 * <li>Quitting the program
 * </ul>
 * <p>
 * If the user chooses the first option, giving the path of 2 already existing
 * files, the program will take both files and print the longest common
 * substring.
 * 
 * In this case the user has 2 choices either search for a string of no more
 * than 30 characters in another of length up to 4 Million character or search
 * between two equal files that contains no more than 20,000 character each or
 * something between these two extremes such as comparing a file with 30,000
 * character with another of 3,000 character. <b>Depending on the size of your
 * heap</b>
 * <p>
 * If the user chooses the second option the program will generate a user
 * specified number of files of a constant length of 500 alphanumeric characters
 * at the user specified paths.
 * <p>
 * After executing any of these 2 commands the program will return to the menu
 * so that the user chooses another option or quit.
 * 
 * The program is Case sensetive
 * 
 * @author Asma Berriri <asma.berriri@telecom-sudparis.eu>
 * @author Mostafa Fateen <mostafafateen@gmail.com>
 * @author Ramzi Sellami <ramzi_sellami@yahoo.com>
 * @author Rebeca Machado <rebeca.machado_garroz@telecom-sudparis.eu>
 * 
 * @version 1.1.0
 * @since 13/2/2015
 */

public class LongestCommonSubstring {

	/**
	 * This is the main of our program in which we do our operations and call
	 * other classes and functions.
	 *
	 * @throws IOException
	 *             if an input or output exception occurred
	 */
	public static void main(String[] args) throws IOException {

		/** A variable in which we put the choice of the user from the menu */
		String choice = null;
		/**
		 * <code>@SuppressWarnings("resource")</code> to suppress warnings
		 * relative to usage of resources of type Closable
		 */
		@SuppressWarnings("resource")
		/** A variable to read from the standard input */
		Scanner scan = new Scanner(System.in);

		do {
			/** Prints the main menu to the user */
			System.out.println("------------------------------------");
			System.out.println("What type of files are you going to use?");
			System.out.println("------------------------------------");
			System.out.println("1) I have already generated my files");
			System.out.println("2) I want to randomly generate my files now");
			System.out.println("3) Quit");
			System.out.println("------------------------------------");

			choice = scan.nextLine();
			switch (choice) {
			/**
			 * <b>The case of the already generated files</b>: Here the user
			 * specifies the path of each file, the program will open both of
			 * them and put them in two different strings <code>str1</code> and
			 * <code>str2</code> by invoking the function {@link getFile()} from
			 * the class <code>FilesReader</code>.
			 * <p>
			 * The program will then call the function {@link lcs()} from the
			 * class <code>LongestCommonSubstringFinder</code> to find the
			 * longest common substring and put it in the local variable
			 * <code>result</code>.
			 * <p>
			 * After that the program will display both files and then the
			 * longest common substring.
			 */
			case "1":
				try {

					System.out
							.println("Please enter the path of the first file: ");
					/** A variable to store the path of the first file */
					String path1 = scan.nextLine();
					/** A variable to store the content of the first file */
					String str1 = FileManager.readFile(path1);

					System.out
							.println("Please enter the path of the second file: ");
					/** A variable to store the path of the second file */
					String path2 = scan.nextLine();

					/** A variable to store the content of the second file */
					String str2 = FileManager.readFile(path2);

					/**
					 * A variable to store the longest common substring between
					 * the 2 files
					 */
					String result = LongestCommonSubstringFinder.findLCS(str1,
							str2);

					System.out.println(str1);
					System.out
							.println("*************************************************");
					System.out.println(str2);
					System.out.println("\nLongest Common Substring : \""
							+ result + "\"");
				} catch (OutOfMemoryError e) {
					System.out
							.println("Out of memory, please extend your heap size or enter smaller files.");
				}
				break;

			/**
			 * <b>The case of generating random files</b>: Here the program
			 * takes from the user the number of files to be generated and calls
			 * the function {@link setFile()} from the class
			 * <code>FilesWriter</code> to do so.
			 */
			case "2":
				System.out.println("How many files do you want to generate?");

				/** A variable to store the number of files to be generated */
				int nbr = 0;
				while (nbr == 0) {
					try {
						nbr = Integer.parseInt(scan.nextLine());
					} catch (NumberFormatException e) {
						System.out.println("Introduce a number bigger than 0.");
					}
				}
				FileManager.writeFile(nbr);
				break;

			/**
			 * <b>The case of Quitting<b>: Here we exit the loop and in turn
			 * exit the program.
			 */
			case "3":
				System.out.println("See you next time!");
				break;

			/**
			 * <b>The default case<b>: in this case we just discard what was
			 * inputted and re-loop to get the correct input.
			 */
			default:
				System.out.println("Error, type 1, 2 or 3!");
				break;

			} // end of the switch case
		} while (!choice.equals("3")); // end of the do..while loop

	}

}

File: C_Version_1.0/src/longest_common_substring.c
/**
 * @file longest_common_substring.c
 * @author Asma Berriri <asma.berriri@telecom-sudparis.eu>
 * @author Mostafa Fateen <mostafa.fateen@telecom-sudparis.eu>
 * @author Rebeca Machado <rebeca.machado_garroz@telecom-sudparis.eu>
 * @author Ramzi Sellami <ramzi.sellami@telecom-sudparis.eu>
 * @version 1.0
 * 
 * Gets two files as input from command line when running the program and 
 * retrieve the longest substring between both in stdout.
 * 
 */

#include <stdio.h>
#include <string.h>

#include "file_manager.h"
#include "longest_common_substring_finder.h"


/**
 * @var file1
 * @brief Saves the name of the first input file.
 */
char *file1;

/**
 * @var file2
 * @brief Saves the name of the second input file.
 */
char *file2;


/**
 * display_result
 * 
 * @brief Receives an array of char with the LCS result and show it in stdout.
 * 
 * @param result Result to be processed and displayed.
 * 
 * The result should be a array of characters holding the longest substring
 * between the two input files.
 * 
 * If the array is not empty then the output will follow this format:
 * 
 * <b>The longest substring between [Input File 1] and [Input File 2] is
 * "[longest substring]".</b>
 *
 * Otherwise, the output would be <b>No common substrings</b>
 * 
 * If the return value is asigned to a variable outside the function, there's no
 * need of allocating memory for it since the function does it. 
 *
 */
void display_result(char *result) {
    
    if (result != NULL) {
        printf("The longest substring between %s and %s is \"%s\".\n",
                file1, file2, result);
    } else {
        printf("No common substrings\n");
    }
}


/**
 * main
 * 
 * @brief Main function.
 * 
 * @param argc Number of parameters received by command line when running.
 * @param argv Value of parameters received by command line when running.
 * @return Integer with value 0 in case of no errors, -1 otherwise.
 * 
 * This main checks the input as it should be two files of extension .txt, then 
 * reads the files and shows the LCS between both.
 *
 */
int main(int argc, char *argv[]) {
    char *s1;
    char *s2;
    char *result;
    char *ext;
    
    if (argc < 3) {
        printf("Usage: longest_common_substring <file1> <file2>\n");
        return(-1);
    }
    
    file1 = argv[1];
    file2 = argv[2];
    
    ext = strrchr(file1, '.');
    if (!ext) {
        fprintf(stderr, "File1 extension should be .txt\n");
        return(-1);
    }
    ext = ext + 1;
    if (strcmp(ext,"txt")) {
        fprintf(stderr, "File1 extension should be .txt\n");
        return(-1);
    }
    
    ext = strrchr(file2, '.');
    if (!ext) {
        fprintf(stderr, "File2 extension should be .txt\n");
        return(-1);
    }
    ext = ext + 1;
    if (strcmp(ext,"txt")) {
        fprintf(stderr, "File2 extension should be .txt\n");
        return(-1);
    }
    
    s1 = read_file(file1);
    if (s1 == NULL) {
        return -1;
    }
    
    s2 = read_file(file2);
    if (s2 == NULL) {
        return -1;
    }
    
    result = find_LCS(s1,s2);
    display_result(result);
    
    return 0;
}